EFS

ryanycoleman opened the pull request:
This PR rebases upstream changes to the gatsby-starter and introduces a basic EFS volume to handle non-database content in a way that's persistent and survives container thrash.

Commit:
add AWS SAM templates for deploying Ghost & Gatsby as serverless infrastructure

Commit:
add secure-headers edge function

Commit:
basic howto template in repo

renovate-bot committed:
Update dependency @tryghost/helpers to v1.1.30

renovate-bot committed:
Update dependency eslint-plugin-react to v7.20.6

renovate-bot committed:
Update dependency @tryghost/helpers-gatsby to v1.0.34

renovate-bot committed:
Update dependency eslint to v7.7.0

endymion1818 committed:
Update index.js

filter data-schema stub out of posts on the index page. Fixes #227

renovate-bot committed:
Update dependency lodash to v4.17.20

renovate-bot committed:
Update dependency eslint to v7.8.1

renovate-bot committed:
Update gatsby monorepo

renovate-bot committed:
Update babel monorepo

fedetibaldo committed:
Ignore environment configurations

Gatsby makes active use of `.env.*` files. It's better to ignore the most common ones. Source: https://www.gatsbyjs.com/docs/environment-variables/#client-side-javascript

renovate-bot committed:
Update gatsby monorepo

aileen committed:
Updated dependencies

renovate-bot committed:
Update dependency sharp to v0.26.0

aileen committed:
Exclude data-stubs from page creation

refs #227

- We should not create pages for any content type with a `data-schema` slug
- This is only a temporary solution until [schema customization](https://github.com/gatsbyjs/gatsby/issues/10856) is implemented

aileen committed:
Allow overwrite of siteUrl with ENV variable

aileen committed:
Exclude data-schema content from sitemap

aileen committed:
Updated gatsby-plugin-advanced-sitemap@1.5.6

renovate-bot committed:
Update gatsby monorepo

aileen committed:
Updated dependencies

aileen committed:
Removed unneded `data-schema` filters

no issue

- With gatsby-source-ghost@4.2.0 we don't need any data-schema stub anymore
- The updated source plugin uses GraphQL Schema using Gatsby Schema Customisation API

renovate-bot committed:
Update gatsby monorepo

renovate-bot committed:
Update Test & linting packages

renovate-bot committed:
Update dependency @tryghost/helpers to v1.1.31

renovate-bot committed:
Update dependency @tryghost/helpers-gatsby to v1.0.35

renovate-bot committed:
Update dependency eslint-plugin-react to v7.21.3

renovate-bot committed:
Update gatsby monorepo

renovate-bot committed:
Lock file maintenance

renovate-bot committed:
Update Test & linting packages

renovate-bot committed:
Update gatsby monorepo

Commit:
working draft of EFS storage instead of RDS Aurora MySQL

Commit:
strip out complicated handling of ghost env data to simplify codebuild builds

Commit:
comment out RSSFeed until I can debug the failure

Commit:
tweaks to clean up gatsby build

Commit:
final tweaks towards a functional efs volume for non-db content

Commit:
uncomment stuff that was broken before rebase

ryanycoleman commented:
@txase I'll handle the weird merge conflict and squash commits on merge. Looking for your impression on the ghostTemplate.yaml based on our conversation this morning. 
https://github.com/stackery/gatsby-starter-ghost/pull/2/files#diff-8d29f0309d096ab569fc271b6fb27162d6de92d843516bfb93ec983d0238c36a

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+        - !GetAtt ghostEFSSecurityGroup.GroupId
+      SubnetId: !Select 
+        - 3 
+        - !Ref DefaultVPCSubnets
```
I think we only want to create ghostMount0 and ghostMount1. There is one default subnet for each AZ in the region this is deployed to, but there are only guaranteed to be two AZs in any region. This is an article showing how many AZs there were in each region at its publication time demonstrating this: https://medium.com/@jaychapel/5-things-you-need-to-know-about-aws-regions-and-availability-zones-60af79b718f9.

ryanycoleman replied:
Interesting. Thanks for the post! 

![Screen Shot 2020-11-01 at 4 09 15 PM](https://user-images.githubusercontent.com/1148878/97820149-9e68b600-1c61-11eb-8796-d1efd7b01701.png)

I've been testing this implementation out and while functional, the ECS service will take longer to become healthy because of task invocations that sit on a subnet without a mount target. I suppose that's fine for this stack, though I should note it in our docs.

txase replied:
I'm leaving comments in a follow up review of the template to address this.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+              awslogs-stream-prefix: logs
+          MountPoints:
+            - 
+              ContainerPath: /var/lib/ghost/content
```
Usually we format things where the above two lines are collapsed into the same line. Just a cosmetic thing, though.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+          Name: 0
+          Environment:
+            - Name: url
+              Value: !Join [ '', [ 'http://', !GetAtt Loadbalancer.DNSName ] ]
```
Out of curiousity, does it need to be able to hit itself? Why is this provided to the container?

ryanycoleman replied:
Certain Ghost functionality depends on `url` which must be the public URL for the CMS. For example, image serving and in-console site previews. https://ghost.org/docs/concepts/config/

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+            # - Name: database__client
+            #   Value: sqlite3
+            # - Name: database__connection__filename
+            #   Value: '/var/lib/ghost/content/data/ghost-efs.db'
```
Should this be cleaned up now?

ryanycoleman replied:
I've refactored all of this to sit on a conditional so that when EnvironmentTagName != 'prod', sqlite will be used instead of Aurora MySQL. I figured that's an easy no-fuss switch for users that will try to do the right thing based on context, but easy enough for users to refactor if they'd like to customize the condition.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+            - Name: DB_ARN
+              Value: !Sub arn:aws:rds:${AWS::Region}:${AWS::AccountId}:cluster:${ghostDB}
+            - Name: DB_ROOT_USER_SECRET_ARN
+              Value: !Ref ghostDBRootUserSecret
```
I suppose you could delete the above two env vars as they don't appear to be used.

ryanycoleman replied:
I believe I included them for debugging purposes early on, bc it's easy to view things through the task definition. But I think you're right that they're dormant.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+            - Name: database__connection__port
+              Value: '3306'
+            - Name: database__connection__database
+              Value: !Ref ghostDB
```
This looks like it's supposed to be the name of the database within the database cluster. The value provided here is the identifier of the cluster itself. If you simply need to provide "a" name, I'd hard code it to something like "ghost".

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+            # https://www.amazontrust.com/repository/AmazonRootCA1.pem
+            # https://ghost.org/docs/concepts/config/
+            - Name: database__connection__ssl__ca
+              Value: "-----BEGIN CERTIFICATE-----\nMIIDQTCCAimgAwIBAgITBmyfz5m/jAo54vB4ikPmljZbyjANBgkqhkiG9w0BAQsF\nADA5MQswCQYDVQQGEwJVUzEPMA0GA1UEChMGQW1hem9uMRkwFwYDVQQDExBBbWF6\nb24gUm9vdCBDQSAxMB4XDTE1MDUyNjAwMDAwMFoXDTM4MDExNzAwMDAwMFowOTEL\nMAkGA1UEBhMCVVMxDzANBgNVBAoTBkFtYXpvbjEZMBcGA1UEAxMQQW1hem9uIFJv\nb3QgQ0EgMTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALJ4gHHKeNXj\nca9HgFB0fW7Y14h29J ...
```
Interesting! It's nice to see SSL CA validation for the MySQL connection :).

To make this more readable, you could make this a multi-line string:

```yaml
Value: |-
  -----BEGIN CERTIFICATE-----
  MIIDQTCCAimgAwIBAgITBmyfz5m/jAo54vB4ikPmljZbyjANBgkqhkiG9w0BAQsF
  ADA5MQswCQYDVQQGEwJVUzEPMA0GA1UEChMGQW1hem9uMRkwFwYDVQQDExBBbWF6
  ...
```
```

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+          EFSVolumeConfiguration:
+            FilesystemId: !GetAtt ghostFS.FileSystemId
+            AuthorizationConfig:
+              IAM: "DISABLED"
```
I've never configured EFS+ECS volumes before. Do you want me to read up and review this configuration?

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+            FilesystemId: !GetAtt ghostFS.FileSystemId
+            AuthorizationConfig:
+              IAM: "DISABLED"
+    DependsOn: ghostAPILogGroup
```
As there's already a !Ref on the ghostAPILogGroup, in the properties, you don't need the DependsOn. (Pay no attention to the one or two places where we do this as we scaffold resources :))

txase replied:
This still can be removed

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+    Type: AWS::ECS::Service
+    Properties:
+      ServiceName: !Sub ghostAPIservice-${EnvironmentTagName}
+      PlatformVersion: 1.4.0
```
I suggest dropping this. It's unlikely future revisions will break things, and they have shipped significant improvements over time. When it's unspecified it uses the latest version by default.

txase replied:
Worth removing still? I didn't see a comment reply.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+      DesiredCount: 2
+      DeploymentConfiguration:
+        MaximumPercent: 200
+        MinimumHealthyPercent: 100
```
This is the default for DeploymentConfiguration, it can be dropped for simplicity.

txase replied:
Worth removing still? I didn't see a comment reply.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+  Loadbalancer:
+    Type: AWS::ElasticLoadBalancingV2::LoadBalancer
+    Properties: 
+      IpAddressType: ipv4
```
I believe this is the default, so you can drop it.

txase replied:
Worth dropping still? I didn't see a comment either way.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+                  - elasticfilesystem:ClientMount
+                  - elasticfilesystem:ClientWrite
+                  - elasticfilesystem:ClientRootAccess
+                Resource: '*'
```
The EFS actions look like dupes that should be dropped

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+                  - elasticfilesystem:ClientMount
+                  - elasticfilesystem:ClientWrite
+                  - elasticfilesystem:ClientRootAccess
+                Resource: '*'
```
Hooray for EFS not supporting resource arns......

I think this could be secured more by using EFS Access Points, but that would continue to increase the scope of this solution.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+              - Effect: Allow
+                Action:
+                  - secretsManager:GetSecretValue
+                Resource: !Sub arn:${AWS::Partition}:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:/${EnvironmentTagName}/*
```
In theory none of the above permissions are actually used by the source code in the container. I think you could drop them, making this a dummy IAM role with no special permissions.

txase replied:
Is this still necessary? I didn't see a comment either way.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+        IpProtocol: tcp
+        CidrIp: 0.0.0.0/0
+        FromPort: 2049
+        ToPort: 2049
```
I think this should be like the ghostDBSecurityGroup, in that it only allows ingress from the ghostAPISecurityGroup.

txase replied:
Was this addressed? I can't tell...

txase replied:
Nevermind, this was addressed!

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,377 @@
+          Subnets: !Ref DefaultVPCSubnets
+          SecurityGroups:
+            - !GetAtt ghostAPISecurityGroup.GroupId
+            - !GetAtt ghostEFSSecurityGroup.GroupId
```
This shouldn't be here. It should only have the ghostAPISecurityGroup.

txase reviewed (commented):
I see a bunch of small improvements for the template, but overall it looks really good!

Commit:
drop MountTarget we cannot assume

Only two AZs are guaranteed in any given region, which means this
template will not work OOTB in some deployment conditions.

https://jaychapel.medium.com/5-things-you-need-to-know-about-aws-regions-and-availability-zones-60af79b718f9

We do not expect a behavior change on EFS mounts from Fargate ECS.

Commit:
change container name to ghost from 0

Commit:
style fix to ContainerDefinitions::MountPoints list

Commit:
support both EFS & Aurora DBs through Condition

Commit:
remove task params DB_ARN and DB_ROOT_USER_SECRET_ARN

Commit:
refactor EFS SG

Commit:
LifecyclePolicies management and linting

Commit:
tweak and codify LBTG health check rules

Commit:
LBTG stickiness, least, with faster deregistration

Commit:
enable EFS encryption at rest and transit

ryanycoleman commented:
The last 10 commits address feedback, create a conditional flip between EFS-backed sqlite or aurora mysql database, and encrypt EFS in transit and at rest. It does not yet implement access points to further scope which resources can mount the EFS volume, even if they possess the default KMS key.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,381 @@
+      NetworkConfiguration:
+        AwsvpcConfiguration:
+          AssignPublicIp: ENABLED
+          Subnets: !Ref DefaultVPCSubnets
```
We fixed the EFS mountpoints by selecting only the first two subnets of DefaultVPCSubnets (because we don't know how many there will be). But if there are more than two subnets (which is common among all the regions), this is what will cause the service to take a while to spin up and become healthy if ECS starts containers in subnets without an EFS mount point (at least, this is what you and I both believe is the case based on the error message you shared in a comment).

What we can do is modify this (and other references to DefaultVPCSubnets) to be:
```yaml
Subnets:
  - !Select 
    - 0
    - !Ref DefaultVPCSubnets
  - !Select 
    - 1
    - !Ref DefaultVPCSubnets
```

We're venturing into more and more manual stack manipulation, which I don't love, but the EFS mount point CFN specification is annoyingly making us do so. Why couldn't they be like every other resource and just let you specify one "mount point" and provide a list of subnets to connect to?

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,381 @@
+      Name: !Sub ghostAPILB-${EnvironmentTagName}
+      SecurityGroups: 
+        - !GetAtt ghostLBSecurityGroup.GroupId
+      Subnets: !Ref DefaultVPCSubnets
```
IIRC, this will only register listeners in subnets where there are targets. If the Fargate service is updated to only run in the first two default VPC subnets, this will end up only registering listeners in those two subnets. But it might not hurt to double check my memory on this. If you see listeners spun up in other subnets that would be a sign this should also be scoped to be an array of only the first two subnets.

txase commented on ghostTemplate.yaml:
```diff
@@ -0,0 +1,381 @@
+      DBSubnetGroupDescription: !Sub
+        - Stack ${StackTagName} Environment ${EnvironmentTagName} Database ${ResourceName} VPC Subnets
+        - ResourceName: ghostDB
+      SubnetIds: !Ref DefaultVPCSubnets
```
Like with the load balancer targets, I think it's fine for this to have all the VPC subnets, but worth considering if it should be scoped to just the first two. I can't remember how aurora cluster DBs are set up.

Actually, IIRC you may be dropping this entirely in favor of ECS. If so, disregard :).

ryanycoleman replied:
Yeah, we should have a look at that too. I'd like your advice on how to properly scope the security-groups for private VPC-links. One's required, but I'm a little confused on its utility.

txase commented on ghostTemplate.yaml:
```diff
@@ -6,6 +6,8 @@ Resources:
       PerformanceMode: generalPurpose
       Encrypted: false
+      LifecyclePolicies:
+        - TransitionToIA: 'AFTER_14_DAYS'
```
Ooo... aren't you fancy :)

txase commented on ghostTemplate.yaml:
```diff
@@ -135,6 +135,13 @@ Resources:
+        - Key: deregistration_delay.timeout_seconds
+          Value: '60'
+        - Key: stickiness.enabled
+          Value: 'true'
```
Does Ghost really need sessions to be sticky? Most modern apps don't...

txase reviewed (commented):
Lookin' good so far!

Commit:
replace alb with http-gw and cloud map (#3)

This commit replaces the load balancer and target group mechanics
with cheaper and more cloud-native HTTP Gateway and Cloud Map service
discovery resources.

However this infrastructure change will cost far less to run (something
like 10x to 2x cheaper per month as a demo site) and represents a cleaner,
more cloud native approach to hybrid container + serverless infrastructure
that will be easier to scale and reason about.

Also:
* clean up VPC SGs and secrets policy
* adjust private namespace for cloudmap to be more reusable
* add ghostProxyRoute to auto-enable integration and cleanup container subnets
* cleanup syntax for ghostProxyRoute Target
* simplify API with openapi definition instead of integration & route resources
* make ECS::Service unique with AWS::StackName variable
* rely on default ECS naming and be explicit about the /ghost route

Commit:
Merge branch 'master' into efs
